### PR TITLE
Added Learning Curve Generation Feature

### PR DESCRIPTION
**Pull request by adibiasio**
#### Summary of Changes
This pull request introduces enhancements to the AutoGluon framework to support the generation and saving of learning curve artifacts during model training. Note that this is a continuation of a pull request on a fork of autogluon: https://github.com/adibiasio/autogluon/pull/5

#### Intended Functionality and Sample Usage
When running the models with the `learning_curves` argument in `predictor.fit`, the predictor will propogate `ag_params` reflecting these specifications downwards to the models, ultimately generating the desired curves. Call `predictor.learning_curves()` to access them. Sample learning_curves() output:

```
import pandas as pd
from autogluon.tabular import TabularPredictor

train_data = pd.read_csv('https://autogluon.s3.amazonaws.com/datasets/Inc/train.csv')
test_data = pd.read_csv('https://autogluon.s3.amazonaws.com/datasets/Inc/test.csv')

predictor = TabularPredictor(label="age", problem_type="regression")
```
```
###### Simple Usage ######
predictor = predictor.fit(
    train_data=train_data,
    test_data=test_data,  # can specify whether to include test_data (does not impact training)
    learning_curves=True, # initiates curve generation with default eval_metric set by predictor
)
```
```
##### Advanced Usage #####
from autogluon.core.metrics import mean_squared_error

curve_params = {
    "metrics": ["r2", "mean_absolute_error", mean_squared_error] # can include names as strings or autgluon Scorers
    "use_error": True # specify metric format (score or error)
}

predictor = predictor.fit(
    train_data=train_data,
    test_data=test_data,
    learning_curves=curve_params, # pass in curve parameters
)
```
```
# access curves
metadata, curves = predictor.learning_curves()

# plot curves
from autogluon.core.learning_curves.plot_curves import plot_curves
plot_curves(curves, "XGBoost", "r2")
```

#### Results
```
# output of predictor.learning_curves()
data = [ # the full data for the project required for simulations
    { # metadata about the autogluon run
        "problem_type": "regression", 
        "eval_metric": "root_mean_squared_error", 
         ...,
         "models": [
             {"model_type": "LGBModel", ...},
             ...,
         ],
    },
    { # all learning curve data generated
        "LGB": [ # model
            ["train", "val", "test"], # list of all evaluation sets included
            ["log_loss", "accuracy"], # list of all metrics included
            [
                [ # log_loss
                    [0.693147, 0.690162, ...], # train
                    [0.693147, 0.690162, ...], # val
                    [0.693147, 0.690162, ...], # test
                ],
                [ # accuracy
                    [0.693147, 0.690162, ...], # train
                    [0.693147, 0.690162, ...], # val
                    [0.693147, 0.690162, ...], # test
                ],
            ]
        ],
        ...,
    }
]
```

<p align="center">
image
</p>

### CODE REVIEW

File: tabular/src/autogluon/tabular/learner/default_learner.py [hunk 1]
```diff
@@ -364,6 +341,59 @@ def set_predefined_weights(self, X, y):
...
+                )
+
+        return X
+
+    def _apply_cleaner_transform(
+        self, X: DataFrame, y_uncleaned: Series, holdout_frac: float | int, holdout_frac_og: float | int, name: str
+    ) -> tuple[DataFrame, Series, Series | None, float | int]:
+        if X is not None and self.label in X.columns:
+            y_og = X[self.label]
+            X = self.cleaner.transform(X)
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/learner/default_learner.py:364]: Add `is_test` argument and the following:

```suggestion
            y_og = X[self.label]
            len_og = len(X) if is_test else None
            X = self.cleaner.transform(X)
            if is_test and len(X) != len_og:
                raise AssertionError(
                    f"{name} cannot have low frequency classes! Please create a GitHub issue if you see this message, as it should never occur."
                )
```

File: core/src/autogluon/core/models/abstract/abstract_model.py [hunk 2]
```diff
@@ -1167,6 +1175,141 @@ def load(cls, path: str, reset_paths: bool = True, verbose: bool = True):
                 model.model = model._compiler.load(path=path)
         return model
 
+    def save_learning_curves(self, metrics: str | List[str], curves: dict[dict[str : List[float]]], path: str = None) -> str:
```
Comment (Innixma) [inline core/src/autogluon/core/models/abstract/abstract_model.py:1178]: The model should mark that it saved the learning curves after calling this method. That way it can be part of the model.get_info output as a boolean flag. For example `self.saved_learning_curves = True`.

File: core/src/autogluon/core/models/abstract/abstract_model.py [hunk 3]
```diff
@@ -1167,6 +1175,141 @@ def load(cls, path: str, reset_paths: bool = True, verbose: bool = True):
...
+            This should generally be a directory path ending with a '/' character (or appropriate path separator value depending on OS).
+            If None, self.path is used.
+            The final curve file is typically saved to os.path.join(path, curves.json).
+
+        Returns
+        -------
+        path : str
+            Path to the saved curves, minus the file name.
+        """
+        if path is None:
```
Comment (Innixma) [inline core/src/autogluon/core/models/abstract/abstract_model.py:1233]: Don't let the save operation occur if learning curves aren't marked as supported.

```python
if not self._get_class_tags().get("supports_learning_curves", False):
    raise AssertionError(...)
```

File: core/src/autogluon/core/models/abstract/abstract_model.py [hunk 4]
```diff
@@ -1167,6 +1175,141 @@ def load(cls, path: str, reset_paths: bool = True, verbose: bool = True):
...
+            The model file is typically located in os.path.join(path, cls.model_file_name).
+
+        Returns
+        -------
+        learning_curves : List
+            Loaded learning curve data.
+        """
+        if not cls._get_class_tags().get("supports_learning_curves", False):
+            raise AssertionError("Attempted to load learning curves from model without learning curve support")
+
```
Comment (Innixma) [inline core/src/autogluon/core/models/abstract/abstract_model.py:1303]:
```python
if not self.saved_learning_curves:
    raise AssertionError(...)
```

File: tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py [hunk 5]
```diff
@@ -350,20 +396,48 @@ def _train_net(self, train_dataset, loss_kwargs, batch_size, num_epochs, epochs_
...
+                        break
+
+                if stop:
+                    break
+
+                # update learning curve
+                for i, metric in enumerate(scorers):
+                    train_curves[metric.name].append(train_metrics[i])
+                    val_curves[metric.name] += [val_metrics[i]] if val_dataset is not None else []
+                    test_curves[metric.name] += [test_metrics[i]] if test_dataset is not None else []
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/tabular_nn/torch/tabular_nn_torch.py:433]: Best to move this to its own method for modularity.

**Review by Innixma — CHANGES_REQUESTED** [review]
Added comments, generally thinks look good besides my comments

File: core/src/autogluon/core/learning_curves/plot_curves.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        # in jupyter environment, simply call function to view graph
+        # note that returning the matplotlib figure object in a jupyter env
+        # will cause graph to appear twice, so set return_fig to False
+        plot_curves(curves, "GBM", "accuracy", return_fig = False)
+
+        # to save figure to path
+        path = "plot.png"
+        fig.savefig(path)
+    """
+    meta_data, model_data = learning_curves
```
Comment (Innixma) [inline core/src/autogluon/core/learning_curves/plot_curves.py:64]: if `meta_data` isn't used, then don't have it as part of the input arguments.

File: core/src/autogluon/core/learning_curves/plot_curves.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+    ax.set_title(f"Learning Curves for {model} using {metric} as Eval Metric")
+    ax.set_xlabel("iterations")
+    ax.set_ylabel(metric)
+    ax.legend()
+    ax.grid(True)
+    plt.show()
+
+    if return_fig:
+        return fig
```
Comment (Innixma) [inline core/src/autogluon/core/learning_curves/plot_curves.py:93]: No need to `if return_fig`, just return it if there is no downsides.

File: core/src/autogluon/core/learning_curves/plot_curves.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        from autogluon.core.learning_curves.plot_curves import plot_curves
+
+        # in jupyter environment, simply call function to view graph
+        # note that returning the matplotlib figure object in a jupyter env
+        # will cause graph to appear twice, so set return_fig to False
+        plot_curves(curves, "GBM", "accuracy", return_fig = False)
+
+        # to save figure to path
+        path = "plot.png"
+        fig.savefig(path)
```
Comment (Innixma) [inline core/src/autogluon/core/learning_curves/plot_curves.py:62]: nit: remove indent

File: core/src/autogluon/core/learning_curves/plot_curves.py (hunk 7 above)
Comment (adibiasio) [reply core/src/autogluon/core/learning_curves/plot_curves.py:93]: I wouldn't say there are no downsides. The reason I implemented the return value this way is that if you use this function in a jupyter environment and you return the figure, then the plot will be displayed twice (once because of the plt methdos and again because of the figure value). I explained this in the docstring by the way. This is why I made it return the figure by default but allow the user to suppress the return functionality (and thus suppress the double plot output). Let me know if there's a better way to address this, or if it's too niche and we just shouldn't address it.

Comment (Innixma) [reply core/src/autogluon/core/learning_curves/plot_curves.py:93]: Fair enough, can keep as is then

File: tabular/src/autogluon/tabular/predictor/predictor.py [hunk 9]
```diff
@@ -4795,44 +4900,116 @@ def _validate_fit_data(
...
-            unlabeled_features = [column for column in unlabeled_data.columns]
-            train_features, unlabeled_features = self._prune_data_features(train_features=train_features, other_features=unlabeled_features, is_labeled=False)
-            train_features = sorted(np.array(train_features))
-            unlabeled_features = sorted(np.array(unlabeled_features))
-            if np.any(train_features != unlabeled_features):
-                raise ValueError(
-                    "Column names must match between training and unlabeled data.\n" "Unlabeled data must have not the label column specified in it.\n"
-                )
-        return train_data, tuning_data, unlabeled_data
+    def _initialize_learning_curve_params(self, learning_curves: dict | bool | None = None, problem_type: str = "") -> dict:
```
[the remark this answers is not in the document]
Comment (adibiasio) [reply tabular/src/autogluon/tabular/predictor/predictor.py:4951]: I use the problem_type parameter here, because self.problem_type has not yet been initialized. Changed default to None though

File: core/src/autogluon/core/models/abstract/abstract_model.py (hunk 4 above)
Comment (adibiasio) [reply core/src/autogluon/core/models/abstract/abstract_model.py:1303]: This is in a classmethod, so no access to self

Comment (Innixma) [reply core/src/autogluon/core/models/abstract/abstract_model.py:1303]: Sounds good, that makes sense

File: tabular/src/autogluon/tabular/models/xgboost/xgboost_model.py [hunk 10]
```diff
@@ -59,9 +59,9 @@ def _get_default_auxiliary_params(self) -> dict:
 
     # Use specialized XGBoost metric if available (fast), otherwise use custom func generator
     def get_eval_metric(self):
-        eval_metric = xgboost_utils.convert_ag_metric_to_xgbm(ag_metric_name=self.stopping_metric.name, problem_type=self.problem_type)
+        eval_metric = xgboost_utils.convert_ag_metric_to_xgbm(ag_metric_name=self.eval_metric.name, problem_type=self.problem_type)
         if eval_metric is None:
-            eval_metric = xgboost_utils.func_generator(metric=self.stopping_metric, problem_type=self.problem_type)
+            eval_metric = xgboost_utils.func_generator(metric=self.eval_metric, problem_type=self.problem_type)
```
Comment (Innixma) [inline tabular/src/autogluon/tabular/models/xgboost/xgboost_model.py:64]: The default should be using the stopping metric, not the eval metric. This code changes default AutoGluon behavior and should be reverted.

**Review by Innixma — COMMENTED** [review]
Added a comment. Also, unit tests seem to be failing now

**Comment by prateekdesai04** [comment]
/benchmark module=tabular preset=tabular_best benchmark=tabular_full time_limit=1h

**Review by Innixma — APPROVED** [review]
AutoMLBenchmark results show no performance degradation, awesome work @adibiasio!

```
AutoGluon_tabular_best_1h8c_2024_08_27 VS all
                                       framework   Winrate    >    <     =  % Loss Reduction  % Loss Reduction (median)  Avg Fit Speed Diff  Avg Inf Speed Diff      rank  loss_rescaled  time_train_s  time_infer_s  time_train_s_rescaled  time_infer_s_rescaled
0         AutoGluon_tabular_best_1h8c_2024_08_27  0.500000    0    0  1032          0.000000                        0.0            0.000000             0.00000  1.495627       0.321672   3426.849174      0.035777               1.008816               1.174533
1  AutoGluon_tabular_best_1h8c_2024_08_27_PR4411  0.504373  340  331   358         -0.070467                        0.0           -0.016223             0.03455  1.504373       0.330418   3461.635666      0.033396               1.025039               1.139983
```
